### ISSUE TITLE
Constants fail in conversion from SQL when too polymorphic

### ISSUE THREAD

**Issue by hesselink** [tomjaguarpaw/haskell-opaleye#151]
If you use a constant in a context where it's not constrained to a type, you get an error when converting it to a Haskell value. For example:

```haskell
{-# LANGUAGE OverloadedStrings #-}

import Data.CaseInsensitive (CI)
import Data.Text (Text)
import Opaleye
import Database.PostgreSQL.Simple (connectPostgreSQL)

test :: IO [CI Text]
test = do
  conn <- connectPostgreSQL ""
  runQuery conn (pure $ constant ("hello" :: CI Text) :: Query (Column PGCitext))
```

This will give:

```
*** Exception: Incompatible {errSQLType = "unknown", errSQLTableOid = Nothing, errSQLField = "result1_1", errHaskellType = "CI Text", errMessage = ""}
```

Sometimes postgres will also default to something instead of giving "unknown". For example, I implemented array literals (would you like me to contribute that back BTW?) and inside array literals, the textual value above will default to 'text' (not 'citext').

It looks to me like the solution would be to always add a type annotation to literals. It could probably be added to the `Constant` class, although that would mean duplication when multiple haskell types map to the same postgres type. In our silk-db-opaleye package we have a type family `PGRep` mapping from Haskell type to postgres type, which we could probably use to remove that duplication.

**Comment by tomjaguarpaw** [comment]
> I implemented array literals (would you like me to contribute that back BTW?)

Yes please!

**Comment by hesselink** [comment]
Cool, I'll open a PR to discuss it further, since there are some rough edges.

**Comment by tomjaguarpaw** [comment]
I agree this is quite annoying.  Adding types to all generated literals is possible, although it will make the generated code somewhat awkward looking.

A short term workaround would be to use `unsafeCast` http://haddock.stackage.org/lts-5.4/opaleye-0.4.2.0/Opaleye-Column.html

**Comment by hesselink** [comment]
Yes, that's what I'm using for now.

**Comment by hesselink** [comment]
Small update since it was mentioned here a few comments up: I sent the array literal generation PR in #154. That includes always casting them with a type class, so it might be relevant to this ticket as well.
